### PR TITLE
test for for-of optimization on arrays and add it for array type anno…

### PR DESCRIPTION
**Pull request by hzoo**
was going to do this earlier but I closed the pr

### CODE REVIEW

**Comment by danez** [comment]
This assumes that people have the correct type annotation. But what if they do what we do? Like having trypanosoma but not verifying them. Then even if the type annotation is Array it could be an object.

**Comment by hzoo** [comment]
Hm good point - would we be ok with that assumption? @jridgewell thoughts?

**Comment by hzoo** [comment]
Linking @Kovensky's suggestion too http://astexplorer.net/#/dh8bycNK1A/2

**Comment by jridgewell** [comment]
I think the biggest improvement here is when we can prove it's an array, so @Kovensky's would work just fine for me.

**Comment by hzoo** [comment]
Updated http://astexplorer.net/#/zoFUOACwxU/3

**Comment by hzoo** [comment]
We could remove the type annotation check I guess and it would just check for binding/array.from

**Comment by jridgewell** [comment]
How about with `arguments`?

**Comment by hzoo** [comment]
@danez I think if you are going to opt-in to types, it's on you to make sure you run flow/ts. So it's a good assumption to make

**Comment by hzoo** [comment]
via @ljharb: `Array.of, new Array(), Array`, `String.prototype.match`, `RegExp.prototype.exec`, and friends

File: packages/babel-plugin-transform-es2015-for-of/test/fixtures/opt/typeannotation/actual.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+for (const y of (b: Array<any>)) {}
```
Comment (hzoo) [inline packages/babel-plugin-transform-es2015-for-of/test/fixtures/opt/typeannotation/actual.js]: Didn't add the type inference yet but we can

**Comment by hzoo** [comment]
I think we can land this as is since I didn't add the type part so it's safe

File: packages/babel-plugin-transform-es2015-for-of/src/index.js [hunk 2]
```diff
@@ -89,16 +94,48 @@ export default function ({ messages, template, types: t }) {
...
+  }
+
+  function optimize(path, right) {
+    if (right.isArrayExpression() || right.isGenericType("Array")) {
+      return replaceWithArray(path);
+    } else if (right.isIdentifier() && right.isPure()) {
+      const binding = path.scope.getBinding(right.node.name);
+      return optimize(path, binding.path.get("init"));
+    } else if (right.isCallExpression() && (
+      isArrayFrom(right.get("callee").node) ||
```
Comment (jridgewell) [inline packages/babel-plugin-transform-es2015-for-of/src/index.js]: I'd rather see this in https://github.com/babel/babel/blob/622672d397805a22ffd8bf47e750e4a871870d41/packages/babel-traverse/src/path/inference/inferers.js#L148-L150.

Comment (hzoo) [inline packages/babel-plugin-transform-es2015-for-of/src/index.js]: I'd be happy to let you finish this up if you want to refactor

Comment (jridgewell) [inline packages/babel-plugin-transform-es2015-for-of/src/index.js]: Let's merge this, I'll add this to https://github.com/babel/babel/pull/5835, and you'll promise to review it. 😉

**Pull request merged**